src/Http/WebUtilities/src/HttpRequestStreamReader.cs

```diff
@@ -310,6 +335,57 @@ public override async Task<int> ReadAsync(char[] buffer, int index, int count)
             return charsRead;
         }
 
+        public override async Task<string> ReadLineAsync()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpRequestStreamReader));
+            }
+
+            StringBuilder sb = new StringBuilder();
+            while (true)
+            {
+                if (_charBufferIndex == _charsRead)
+                {
+                    if (await ReadIntoBufferAsync() == 0)
+                    {
+                        break;  // reached EOF, we need to return null if we were at EOF from the beginning
+                    }
+                }
+
+                var ch = _charBuffer[_charBufferIndex++];
```

Comment:
Going character by character seems like it could be inefficient here. I think we should try to benchmark this. Could you make a Span or Memory<char> of _charbuffer and call FirstIndexOf('\n'), then check the character before for the \r case. That way, you aren't appending a character at a time as you can slice the necessary string.

Just a hunch, and honestly this may be faster.

Reply:
This implementation is based on WriteLine on the base class, as the goal was just to make it async, avoiding sync over async.

However, this is a good suggestion and I started to play with it. There are a few edge cases to handle so it's not enough to do a quick benchmark and I need a correct implementation of the whole logic to properly benchmark (and a bit worried it could be more complex).

However there is a big potential for saving bypassing the StringBuilder when the new line is found on the first pass, which would be also beneficial for the sync version, so I will carry on pursuing this solution.